hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/test/StreamSourceLong.java

```diff
@@ -0,0 +1,155 @@
+import com.hazelcast.jet.core.AppendableTraverser;
+import com.hazelcast.jet.core.EventTimePolicy;
+import com.hazelcast.jet.core.Watermark;
+
+import javax.annotation.Nonnull;
+
+import static com.hazelcast.jet.impl.JetEvent.jetEvent;
+import static java.util.concurrent.TimeUnit.HOURS;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Represents a StreamSource containing Long values which can be used
+ * for performance pipeline testing and development.
+ *
+ * @since 4.3
+ */
+public class StreamSourceLong extends AbstractProcessor {
+
+    private static final long SOURCE_THROUGHPUT_REPORTING_PERIOD_SECONDS = 10;
+    private static final long SIMPLE_TIME_SPAN_MILLIS = HOURS.toMillis(3);
+
+    private static final long REPORT_PERIOD_NANOS = SECONDS.toNanos(SOURCE_THROUGHPUT_REPORTING_PERIOD_SECONDS);
+    private static final long HICCUP_REPORT_THRESHOLD_MILLIS = 10;
+    private final long nanoTimeMillisToCurrentTimeMillis = determineTimeOffset();
+    private final long startTime;
+    private final long itemsPerSecond;
+    private final boolean isReportingThroughput;
+    private final long wmGranularity;
+    private final long wmOffset;
+    private long globalProcessorIndex;
+    private long totalParallelism;
+    private long emitPeriod;
+
+    private final AppendableTraverser<Object> traverser = new AppendableTraverser<>(2);
+    private long emitSchedule;
+    private long lastReport;
+    private long counterAtLastReport;
+    private long lastCallNanos;
+    private long counter;
+    private long lastEmittedWm;
+    private long nowNanos;
+
+    /**
+     * Creates a stream source for Long values that is supposed to be used for performance testing purposes.
+     *
+     * @since 4.3
+     */
+    @Nonnull
+    public StreamSourceLong(
+            long startTime,
+            long itemsPerSecond,
+            EventTimePolicy<? super Long> eventTimePolicy,
+            boolean shouldReportThroughput
+    ) {
+        this.wmGranularity = eventTimePolicy.watermarkThrottlingFrameSize();
+        this.wmOffset = eventTimePolicy.watermarkThrottlingFrameOffset();
+        this.startTime = MILLISECONDS.toNanos(startTime + nanoTimeMillisToCurrentTimeMillis);
+        this.itemsPerSecond = itemsPerSecond;
+        this.isReportingThroughput = shouldReportThroughput;
+    }
+
+    @Override
+    protected void init(Context context) {
+        totalParallelism = context.totalParallelism();
+        globalProcessorIndex = context.globalProcessorIndex();
+        emitPeriod = SECONDS.toNanos(1) * totalParallelism / itemsPerSecond;
+        lastCallNanos = lastReport = emitSchedule =
+                startTime + SECONDS.toNanos(1) * globalProcessorIndex / itemsPerSecond;
+    }
+
+    @Override
+    public boolean complete() {
+        nowNanos = System.nanoTime();
+        detectAndReportHiccup();
```

Comment:
You can reorder these two calls so that `emitEvents()` is first, then `detectAndReportHiccup()`. It is a bit cleaner that way.